src/sympc/tensor/additive_shared.py

```diff
@@ -18,20 +18,59 @@ def __init__(self, secret=None, shares=None, session=None):
         self.session = session
         self.shape = None
 
-        conf = session.config
-        self.fp_encoder = FixedPointEncoder(base=conf.enc_base, precision=conf.enc_precision)
-
         if secret is not None:
+            secret, shape, is_remote_secret = AdditiveSharingTensor.sanity_checks(secret, shape, session)
             parties = session.parties
-            secret = self.fp_encoder.encode(secret)
-            self.shape = secret.shape
-            shares = AdditiveSharingTensor.generate_shares(secret, session)
-            self.shares = []
-            for share, party in zip(shares, parties):
-                self.shares.append(share.send(party))
+            self.shape = shape
+
+            if is_remote_secret:
+                # If the secret is remote we use PRZS (Pseudo-Random-Zero Shares) and the
+                # party that holds the secret will add it to it's share
+                self.shares = AdditiveSharingTensor.generate_przs(self.shape, self.session)
+                for i, share in enumerate(self.shares):
+                    if share.client == secret.client:
+                        self.shares[i] = self.shares[i] + secret
+            else:
+                self.shares = []
+                shares = AdditiveSharingTensor.generate_shares(secret, self.session)
+                for share, party in zip(shares, self.session.parties):
+                    self.shares.append(share.send(party))
+
         elif shares is not None:
             self.shares = shares
 
+
+    @staticmethod
+    def sanity_checks(secret, shape, session):
+        is_remote_secret = False
+
+        if "Pointer" in type(secret).__name__:
```

Comment:
#HackAlert ;)

Reply:
Yep - I need to move this into a new function